clang-tools-extra/clang-tidy/hooks/GuardCallNonConstCheck.h

```diff
@@ -0,0 +1,33 @@
+//===--- GuardCallNonConstCheck.h - clang-tidy ------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_HOOKS_GUARDCALLNONCONSTCHECK_H
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_HOOKS_GUARDCALLNONCONSTCHECK_H
+
+#include "../ClangTidyCheck.h"
+
+namespace clang {
+namespace tidy {
+namespace hooks {
+
+/// Only const integers or literals can be used as an argument in loop GUARD call.
```

Comment:
That isn't really true - `_g(1,1+1)` works just fine (and is in fact used by the GUARD implementation). The actual rule is that the second argument must be a compile-time constant - and it can be checked, by matching a generic expression, then trying to get its compile-time value (by calling `getIntegerConstantExpr` on it).

Also, what about the first argument? That also must be a compile-time constant, but in addition, it must be unique among all the calls to `_g` is a single hook. Could this check be extended to also warn about that?

Reply:
Ok, will update that. Not sure if it will be possible to check the uniqueness of the first argument, but I'll try to extend it.